internal/controller/compaction/reconciler.go

```diff
@@ -359,6 +353,129 @@ func (r *Reconciler) createCompactionJob(ctx context.Context, logger logr.Logger
+		metricJobsTotal.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededUnknown, druidmetrics.EtcdNamespace: etcd.Namespace}).Inc()
+		metricJobDurationSeconds.With(prometheus.Labels{druidmetrics.LabelSucceeded: druidmetrics.ValueSucceededUnknown, druidmetrics.EtcdNamespace: etcd.Namespace}).Observe(time.Since(job.Status.StartTime.Time).Seconds())
+	}
+
+	if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+		return fmt.Errorf("error while deleting failed compaction job: %v", err)
+	}
+	return nil
+}
+
+func handleBackoffLimitExceeded(ctx context.Context, r *Reconciler, logger logr.Logger, job *batchv1.Job, etcd *druidv1alpha1.Etcd) error {
+	logger.Info("Job has been completed due to backoffLimitExceeded", "namespace", job.Namespace, "name", job.Name)
+	pod, err := getPodForJob(ctx, r.Client, logger, job)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Couldn't fetch pods for job", "namespace", job.Namespace, "name", job.Name)
+			return fmt.Errorf("error while fetching pod for job: %v", err)
+		}
+		logger.Info("No pods found for job", "namespace", job.Namespace, "name", job.Name)
+	}
+	podFailureReason, lastTransitionTime := getPodFailureReasonAndLastTransitionTime(pod)
+	handlePodFailureReason(logger, pod, etcd, podFailureReason, lastTransitionTime, job.Status.StartTime.Time)
+	return nil
+}
+
+func handlePodFailureReason(logger logr.Logger, pod *v1.Pod, etcd *druidv1alpha1.Etcd, reason podFailureReason, lastTransitionTime time.Time, jobStartTime time.Time) {
+	var labelValue string
+	switch reason {
+	case podReasonPreemptionByScheduler:
+		logger.Info("Pod has been preempted by the scheduler", "namespace", pod.Namespace, "name", pod.Name)
+		labelValue = druidmetrics.ValueSucceededPreempted
+	case podReasonDeletionByTaintManager, podReasonEvictionByEvictionAPI, podReasonTerminationByKubelet:
+		logger.Info("Pod has been evicted", "namespace", pod.Namespace, "name", pod.Name, "reason", reason)
+		labelValue = druidmetrics.ValueSucceededEvicted
+	case podReasonProcessFailure:
+		logger.Info("Pod has failed due to process failure", "namespace", pod.Namespace, "name", pod.Name)
+		labelValue = druidmetrics.ValueSucceededFalse
+	default:
+		logger.Info("Pod has failed due to unknown reason", "namespace", pod.Namespace, "name", pod.Name)
+		labelValue = druidmetrics.ValueSucceededUnknown
+	}
+
+	metricJobsTotal.With(prometheus.Labels{druidmetrics.LabelSucceeded: labelValue, druidmetrics.EtcdNamespace: etcd.Namespace}).Inc()
+	if !lastTransitionTime.IsZero() {
+		metricJobDurationSeconds.With(prometheus.Labels{druidmetrics.LabelSucceeded: labelValue, druidmetrics.EtcdNamespace: etcd.Namespace}).Observe(lastTransitionTime.Sub(jobStartTime).Seconds())
+	}
+}
+
+// getJobCompletionStatusAndReason returns whether the job is successful or not and the reason for the completion.
+func getJobCompletionStatusAndReason(job *batchv1.Job) (bool, string) {
```

Comment:
The name of the method indicates that you will return a Job completion status and a reason but what you return is  a bool and a reason. In the method you only handle successful and failed completion. If none of the conditions are set then you by default return false indicating its a failure but that is not true. It could also be suspended.